RFCs/FS-1090-Generic-struct-type-whose-fields-are-all-unmanaged-types-is-unmanaged.md

```diff
@@ -0,0 +1,198 @@
+# F# RFC FS-1088 - (Generic struct type whose fields are all unmanaged types is unmanaged)
+
+The design suggestion [Generic struct type whose fields are all unmanaged types is unmanaged](https://github.com/fsharp/fslang-suggestions/issues/692) has been marked "approved in principle".
+This RFC covers the detailed proposal for this suggestion.
+
+* [x] Approved in principle
+* [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/692)
+* [ ] Implementation: [In progress](https://github.com/dotnet/fsharp/pull/6064)
+
+# Summary
+
+[summary]: #summary
+
+Allow generic structs to be `unmanaged` at construction if all fields are `unmanaged` types.
+
+# Motivation
+
+[motivation]: #motivation
+
+* Make it easier to author low level interop code.
+* Improve interoperability with C#.
+
+# Detailed design
+
+[design]: #detailed-design
+
+Currently, it is not possible to have an `unmanaged` constraint working on generic struct types.
+
+Consider following code:
+
+```fsharp
+[<Struct>]
+type Test<'T when 'T: unmanaged> =
+        val element: 'T
+
+let works = Test<int>()
+let error = Test<Test<int>>()
+
```

Comment:
I recommend the code sample also includes an example of a custom struct that does work.

```suggestion
[<Struct>]
type MyStruct(x: int, y: int) =
    member _.X = x
    member _.Y = y
    
[<Struct>]
type MyStructGeneric<'T when 'T: unmanaged>(x: 'T, y: 'T) =
    member _.X = x
    member _.Y = y

[<Struct>]
type Test<'T when 'T: unmanaged> =
    val element: 'T
        
let works = Test<int>()
let works2 = Test<MyStruct>()

// error FS0001: A generic construct requires that the type 'MyStructGeneric<int>' is an unmanaged type
let error = Test<MyStructGeneric<int>>()
```

Reply:
This also means that this won't work either:

```fsharp
[<Struct>]
type MyStruct(x: byref<int>, y: byref<int>) =
    member _.X = x
    member _.Y = y
```

Which feels like a big gap in the feature area.

Reply:
Nor this

```fsharp
[<Struct>]
type MyStruct(x: Span<int>, y: Span<int>) =
    member _.X = x
    member _.Y = y
```

Nor this:

```fsharp
let fails = Test<Span<int>>()
```